feat(tracing): propagate gcp trace header in tracing plugins

jonabc opened the pull request:
<!--
NOTE: Please read the CONTRIBUTING.md guidelines before submitting your patch,
and ensure you followed them all:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributing
-->

### Summary

<!--- Why is this change required? What problem does it solve? -->

This change parses and propagates the `X-Trace-Context` header that Google provides for calls in Google Cloud Projects (GCP).  Google [recommends using the OpenTelemetry tracing standard](https://cloud.google.com/trace/docs/setup#when-to-instrument) when instrumenting applications, and this provides handling of the header at the API gateway layer 👍 

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [x] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - https://github.com/Kong/docs.konghq.com/pull/5829

### Full changelog

- **OpenTelemetry**: Support GCP X-Cloud-Trace-Context header
  The field `header_type` now accepts the value `gcp` to propagate the
  Google Cloud trace header

### Issue reference

<!--- If it fixes an open issue, please link to the issue here. -->
<!--Fix #_[issue number]_-->

I didn't see an open issue for this, but I'm happy to open an issue and link it here if wanted.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=11254) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=11254) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=11254) it.</sub>

jonabc commented:
@hanshuebner :wave: I saw you gave a review on a similar PR for [propagating AWS' xray trace headers](https://github.com/Kong/kong/pull/11075), would you mind taking a look here as well?  Sorry for the ping, this has been sitting around a little while and I'd love to have this included in Kong if possible 🙏

hanshuebner commented:
@samugi Can you have a look?  This seems to be right up your alley.

samugi commented on kong/tracing/propagation.lua:
```diff
@@ -397,6 +411,63 @@ local function parse_aws_headers(aws_header)
+
+  -- GCP header pattern is lenient on field existence and length
+  -- coalesce empty values to nil for easier validation
+  if trace_id == "" then
+    trace_id = nil
+  end
```
Nitpick: you **could** do
`trace_id = trace_id ~= "" and trace_id`

and later check for `falsy` instead of `nil`:
`if not trace_id or ...`

same for the other fields.
But it's really just personal preference, feel free to disregard this one.

samugi commented on kong/tracing/propagation.lua:
```diff
@@ -623,6 +701,13 @@ local function set(conf_header_type, found_header_type, proxy_span, conf_default
+  if conf_header_type == "gcp" or found_header_type == "gcp" then
+    set_header("x-cloud-trace-context", fmt("%s/%s;o=%s", 
+      to_gcp_trace_id(to_hex(proxy_span.trace_id)),
+      to_hex(proxy_span.span_id),
```
similarly to the comment above, span_id should be set as `from_binary(proxy_span.span_id):to_dec()` here. This might suggest a missing test scenario.
```suggestion
      from_binary(proxy_span.span_id):to_dec(),
```

samugi commented on kong/tracing/propagation.lua:
```diff
@@ -397,6 +411,63 @@ local function parse_aws_headers(aws_header)
+  if span_id == nil or #span_id > GCP_SPAN_ID_MAX_LEN then
+    table.insert(invalid_values, "span ID")
+  else
+    -- the span id is a number between 1 and 20 digits long 
+    -- the propagation code expects span ids to be stored as hex strings.
+    -- left pad a 0 if the id has an odd number of digits for a valid hex value
+    if #span_id % 2 ~= 0 then
+      span_id = '0' .. span_id
+    end
+
+    span_id = from_hex(span_id)
```
```suggestion

    span_id = from_dec(span_id):to_binary()
```
the span id is actually stored as binary data (from_hex converts a hex string to binary), in this case the span_id is represented as an unsigned integer so it should be returned as the output of: `from_dec(span_id):to_binary()`.
This may suggest a missing test case that ensures the span_id is treated correctly.

samugi reviewed (changes requested):
Thank you so much for your contribution!

I gave this a first pass, I'll be looking into the tests next but I have already left some comments that might suggest at least one missing test case.

jonabc commented:
@samugi thanks for the feedback!  I think I've addressed everything you brought up in https://github.com/Kong/kong/pull/11254/commits/75344f1fd48e342d19929283c5b624a947646edb (force-pushed to rebase on top of latest kong master).  Would you mind taking another look?

jonabc commented on spec/01-unit/26-tracing/02-propagation_spec.lua:
```diff
@@ -799,6 +899,15 @@ describe("propagation.set", function()
       )
     }
 
+    local gcp_headers 
+    if tonumber(span_id, 10) then
+      -- strip leading 0s in expected output 
+      gcp_headers = {["x-cloud-trace-context"] = gcp_trace_id .. "/" .. tonumber(span_id, 10) .. ";o=1"}
+    else
+      -- hex values are not valid span id inputs, translate to decimal
+      gcp_headers = {["x-cloud-trace-context"] = gcp_trace_id .. "/" .. openssl_bignumber.from_hex(span_id):to_dec() .. ";o=1"}
+    end
```
I needed to detect whether the span id was a decimal or hex value because the span id part of a gcp trace header is written as a decimal value.  using `tonumber(span_id, 16)` on the hex value was resulting in a span_id number like `1.844674407371e+19` instead of `18446744073709551603`, which is why I ended up using the openssl library to do the conversion.

I _think_ this is an ok way to handle this and all tests were passing locally 🤷 .  As I mentioned though I'm pretty new to Lua, please let me know if there's a better way to handle this!

samugi replied:
that's **great**, good job!
You can save a couple of lines though and consider all span IDs to be hex, since span_id == "0000000000000003" == 0x3 the result will be the same.

```suggestion
    -- hex values are not valid span id inputs, translate to decimal
    local gcp_headers = {["x-cloud-trace-context"] = gcp_trace_id .. "/" .. openssl_bignumber.from_hex(span_id):to_dec() .. ";o=1"}
```

jonabc replied:
oh 🤦 yeah good catch.  I've included that updated in the latest push

jonabc commented on kong/tracing/propagation.lua:
```diff
@@ -397,6 +412,50 @@ local function parse_aws_headers(aws_header)
+
+  if not span_id or tonumber(span_id, 10) > GCP_SPAN_ID_MAX_VALUE  then
+    table.insert(invalid_values, "span ID")
+  else
```
❤️ I didn't know about ngx.re.match, thank you!  I wanted to use a regex from the start to match the logic from that propagator.  that's where the 20 digit max length in the previous iteration came from 👍   Using the regex hugely simplifies the logic, updated in https://github.com/Kong/kong/pull/11254/commits/75bc1d225042663ae10535afa4595bc14ab559ac

samugi replied:
awesome, thanks for adding a test as well

samugi commented on kong/tracing/propagation.lua:
```diff
@@ -397,6 +411,32 @@ local function parse_aws_headers(aws_header)
+    return nil, nil, nil
+  end
+
+  local match, err = ngx.re.match(gcp_header, GCP_TRACECONTEXT_REGEX)
```
```suggestion
  local match, err = ngx.re.match(gcp_header, GCP_TRACECONTEXT_REGEX, 'jo')
```
For better performance, see the options section of [ngx.re.match](https://github.com/openresty/lua-nginx-module#ngxrematch)

samugi commented on CHANGELOG.md:
```diff
@@ -116,6 +116,10 @@
 - **Ip-Restriction**: Add TCP support to the plugin.
   Thanks [@scrudge](https://github.com/scrudge) for contributing this change.
   [#10245](https://github.com/Kong/kong/pull/10245)
+- **OpenTelemetry**: Support GCP X-Cloud-Trace-Context header
```
```suggestion
- **OpenTelemetry** and **Zipkin**: Support GCP X-Cloud-Trace-Context header
```

jonabc commented:
@samugi thanks again!  Updated (+ rebase on latest `Kong:master`) in https://github.com/Kong/kong/pull/11254/commits/e1b04fb49f63ba1fe9de809a541235f14a74aea6.  I also moved the changelog entry to `unreleased`, it looks like it was still set in `3.4.0`.  

Do you have any estimates on when 3.5.0 will be released, assuming this will go out with the next feature release?

samugi commented:
> assuming this will go out with the next feature release

@jonabc That is correct. I don't have a date to share yet but I expect 3.5.0 to be out roughly in 2-3 months considering 3.4.0 was just released. Again thanks so much for your contribution!

outsinre commented:
Is this in EE?

samugi commented:
It is now @outsinre
